README
QCADesigner README Notes
============================

If you would like to run QCADesigner from the source tree,
that is, you do not want to perform "make install", you
may do so by switching to the src/ subdirectory and
running ./QCADesigner .

If you would like to install QCADesigner to a path other
than the default you can run configure with the --prefix
option and specify the path. For example if you are
configuring QCADesigner for the first time, you can run:

./autogen.sh --prefix=<Your path>
./configure
make
make install

"ls <Your path>" should show a "bin" and "share" directory
that contain all the files that QCADesigner has installed.
You can then use the script provided above to run
QCADesigner.


ChangeLog
2006-08-12 17:45  nix

	* src/objects/QCADCell.c: When scaling all cells in the design,
	  make sure their properties reflect the new sizes.

2006-07-21 01:44  nix

	* docs/reference/Makefile.am, docs/reference/QCADesigner-docs.sgml,
	  src/objects/QCADLayer.c, src/objects/QCADLayersCombo.c,
	  src/objects/QCADLayersCombo.h,
	  src/objects/QCADPropertyUIDouble.c,
	  src/objects/QCADPropertyUIInt.c,
	  src/objects/QCADPropertyUIText.c: More gtk-doc-umentation.

2006-07-20 17:13  nix

	* docs/reference/Makefile.am,
	  docs/reference/QCADesigner-sections.txt,
	  docs/reference/QCADesigner.types, src/Makefile.am: Do not add .c
	  files to the list of files to link into a gtkdoc-scanner.

2006-07-19 09:38  nix

	* docs/manual.src/create_import_block.html: Corrected duplicate
	  duplicate.

2006-07-19 09:17  nix

	* docs/manual.src/hacking.html: More clarity as to where CVS
	  download instructions are located.

2006-07-19 09:12  nix

	* docs/manual.src/installation.html: QCADesigner package metaname
	  fixups.

2006-07-19 09:08  nix

	* docs/manual.src/installation.html: Typo.

2006-07-17 06:15  nix

	* src/vector_table.c: Ignore NULLs in VectorTable_fill.

2006-07-17 06:14  nix

	* src/: main_batch_sim.c, vector_table.c: Use VectorTable_fill
	  whenever starting a new simulation.

2006-07-17 06:01  nix

	* src/main_batch_sim.c: Include decision to run
	  EXHAUSTIVE_VERIFICATION/VECTOR_TABLE in printf.

2006-07-17 02:28  nix

	* src/main_batch_sim.c: When there's a vector table, set the
	  simulation type to VECTOR_TABLE.

2006-07-11 04:25  nix

	* po/QCADesigner.pot, po/de.po, po/fr.po, po/hu.po, po/it.po,
	  po/pl.po, po/ro.po, src/Makefile.am, src/main_batch_sim.c: Vector
	  table support for batch_sim.

2006-07-11 04:01  nix

	* src/: Makefile.inc, graph_dialog_interface.c,
	  vector_table_options_dialog_callbacks.c,
	  vector_table_options_dialog_interface.c,
	  objects/QCADScrolledWindow.c, objects/QCADScrolledWindow.h:
	  Removed QCADScrolledWindow.

2006-07-10 01:51  nix

	* src/: callbacks.c, interface.c: - Minor cosmetic tweaks.  -
	  Disable "Start Simulation" while simulation is going on.

2006-06-30 20:44  nix

	* src/: graph_dialog.c, graph_dialog_callbacks.c,
	  graph_dialog_data.c, graph_dialog_data.h,
	  graph_dialog_interface.c, graph_dialog_interface.h: Fixed graph
	  dialog resize bug.

2006-06-30 20:05  nix

	* src/selection_undo.c: Added some comments.

2006-06-11 23:31  nix

	* src/: callbacks.c, selection_undo.c, selection_undo.h: Use
	  design_selection_objects_foreach when ndoing/redoing
	  in(de)crement clock.

2006-06-11 22:21  nix

	* src/: selection_renderer.c, selection_renderer.h: Added some doc
	  to selection_renderer.

2006-06-11 22:13  nix

	* src/: selection_undo.c, actions/select.c,
	  objects/QCADDesignObject.c, objects/QCADLayer.c: [no log message]

2006-06-02 02:18  nix

	* src/callbacks.c: Fixed rotation problem. In action button
	  clicked, replace gtk_toggle_button_get_active with g_object_get
	  (G_OBJECT (widget), "active", ...) That way, it's
	  type-independent.

2006-05-31 00:45  nix

	* src/actions/select.c: Make Ctrl+click deselect object if
	  hit-tested.

2006-05-31 00:01  nix

	* src/objects/QCADDesignObject.c: Fixing Ctrl+Select.

2006-05-28 20:37  nix

	* docs/reference/Makefile.am,
	  docs/reference/QCADesigner-sections.txt, src/exp_pixmap.c,
	  src/exp_pixmap.h, src/actions/select.c, src/objects/QCADLabel.c,
	  src/objects/QCADLabel.h, src/objects/QCADToggleToolButton.c:
	  gtk-doc mods.

2006-05-26 15:47  nix

	* src/objects/QCADDesignObject.c: Fixed(?) overlap of exactly
	  adjacent cells.

2006-05-22 21:03  nix

	* src/objects/: QCADUndoEntry.c, QCADUndoEntry.h: Added gtk-doc to
	  QCADUndoEntry.

2006-05-22 20:00  nix

	* src/objects/QCADUndoEntry.c: Trying to get QCADUndoEntry::apply
	  working ...

2006-05-15 22:36  nix

	* src/: graph_dialog.c, graph_dialog_callbacks.c,
	  graph_dialog_callbacks.h, graph_dialog_data.c,
	  graph_dialog_data.h, graph_dialog_interface.c,
	  graph_dialog_interface.h: Moved to GtkSizeGroup from my hacky
	  make-all-graphs-same-cy solution.

2006-05-13 21:20  nix

	* src/: bistable_simulation.c, design.c, design.h: Removed
	  REDUCE_DEREF from bistable sim.

2006-05-04 13:00  nix

	* docs/reference/QCADesigner-sections.txt, src/objects/QCADCell.c:
	  Fixed the cell to display its polarization when set to FIXED.

2006-04-30 19:55  nix

	* src/: print.c, objects/QCADDesignObject.c,
	  objects/QCADRectangleElectrode.c: Express "nmx" and "nmy" in
	  terms of "nm". That way, the scale can be adjusted with only one
	  modification.

2006-04-30 16:16  nix

	* src/interface.c: Added "Copy" menu item to make Ctrl+C work.

2006-04-30 15:54  nix

	* src/translate_selection_dialog.c: Fixed minor segfaults (forgot
	  NULL g_object_new terminator) in the translate selection dialog.

2006-04-30 14:58  nix

	* src/: design.c, design.h, objects/QCADCell.c, objects/QCADCell.h:
	  Commented out stuff related to setting host names on cells.

2006-04-29 14:10  nix

	* src/: callbacks.c, design.c, objects/QCADCell.c: Removed
	  "cell-function-changed" signal. "notify::function" is sufficient.

2006-04-28 02:37  nix

	* ChangeLog: cvs2cl-ized the ChangeLog. Looks great!

2006-04-28 02:03  nix

	* src/: callbacks.c, design.c, objects/QCADCell.c,
	  objects/QCADCell.h: Removed qcad_cell_set_display_mode. Use
	  g_object_set (G_OBJECT (cell), "mode", new_mode, NULL) ;

2006-04-27 14:40  nix

	* docs/reference/QCADesigner-sections.txt,
	  src/objects/QCADDesignObject.c, src/objects/QCADLabel.c:
	  reference changes + QCADLabel get_PostScript_instance check.

2006-04-26 14:36  nix

	* src/: print.c, actions/select.c, objects/QCADCell.c,
	  objects/QCADDesignObject.c, objects/QCADDesignObject.h: Converted
	  PostScript_preamble to a class function.

2006-04-23 22:19  nix

	* src/main_combo_test.c: Minot changes to the combo test.

2006-04-23 03:10  nix

	* docs/reference/QCADesigner-sections.txt, src/objects/QCADLayer.c,

src/print.c
//////////////////////////////////////////////////////////
// QCADesigner                                          //
// Copyright 2002 Konrad Walus                          //
// All Rights Reserved                                  //
// Author: Konrad Walus                                 //
// Email: qcadesigner@gmail.com                         //
// WEB: http://qcadesigner.ca/                          //
//////////////////////////////////////////////////////////
//******************************************************//
//*********** PLEASE DO NOT REFORMAT THIS CODE *********//
//******************************************************//
// If your editor wraps long lines disable it or don't  //
// save the core files that way. Any independent files  //
// you generate format as you wish.                     //
//////////////////////////////////////////////////////////
// Please use complete names in variables and fucntions //
// This will reduce ramp up time for new people trying  //
// to contribute to the project.                        //
//////////////////////////////////////////////////////////
// This file was contributed by Gabriel Schulhof        //
// (schulhof@atips.ca).                                 //
//////////////////////////////////////////////////////////
// Contents:                                            //
//                                                      //
// This file implements a PostScript printer for        //
// QCADesigner designs.                                 //
// Completion Date: June 2003                           //
//                                                      //
//////////////////////////////////////////////////////////

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "print.h"
#include "design.h"
#include "print_util.h"
#include "objects/QCADLayer.h"

#define DBG_P(s)

static GList **PlaceObjectsOnPages (DESIGN *design, print_design_OP *pPO, double cxPageNm, double cyPageNm, double dxMinNm, double dyMinNm) ;

void PrintPages (FILE *pfile, print_design_OP *pPO, GList **lstPages, double dEffPageCYPts, double dxMinNm, double dyMinNm, double dxDiffMinNm, double dyDiffMinNm) ;

static void PrintObjectPreamble (FILE *pfile, GList **lstPages, int icPages) ;

void print_world (print_design_OP *pPO, DESIGN *design)
  {
  FILE *pfile = NULL ;
  GList
    *lstItr = NULL,
    **lstPages = NULL ; // 2D array of GList * (cxPages wide and cyPages long)
  gboolean *pbPrint = NULL, bPrintThisLayer = FALSE, bFirstLayer = TRUE ;
  double dxMinNm = 0, dyMinNm = 0, dxMaxNm = 0, dyMaxNm = 0, cxNm = 0, cyNm = 0,
    dxDiffMinNm = 0, dyDiffMinNm = 0,
    dEffPageCXPts = pPO->po.dPaperCX - pPO->po.dLMargin - pPO->po.dRMargin,
    dEffPageCYPts = pPO->po.dPaperCY - pPO->po.dTMargin - pPO->po.dBMargin ;
  WorldRectangle extentsNm = {0.0} ;
  char doubles[1][G_ASCII_DTOSTR_BUF_SIZE] = {""} ;

  DBG_P (fprintf (stderr, "print_world:Entering\n")) ;

  // Get the size of the world in nanos
  for (lstItr = design->lstLayers ; lstItr != NULL ; lstItr = lstItr->next)
    {
    if (!(bPrintThisLayer = (NULL == (pbPrint = g_object_get_data (G_OBJECT (lstItr->data), PRINT_LAYER_KEY)))))
      bPrintThisLayer = (*pbPrint) ;

    if (bPrintThisLayer)
      {
      qcad_layer_get_extents ((QCAD_LAYER (lstItr->data)), &extentsNm, FALSE) ;
      if (bFirstLayer)
        {
        dxMinNm = extentsNm.xWorld ;
        dyMinNm = extentsNm.yWorld ;
        dxMaxNm = extentsNm.cxWorld + extentsNm.xWorld ;
        dyMaxNm = extentsNm.cyWorld + extentsNm.yWorld ;
        bFirstLayer = FALSE ;
        }
      else
        {
        dxMinNm = MIN (extentsNm.xWorld, dxMinNm) ;
        dyMinNm = MIN (extentsNm.yWorld, dyMinNm) ;
        dxMaxNm = MAX (extentsNm.xWorld + extentsNm.cxWorld, dxMaxNm) ;
        dyMaxNm = MAX (extentsNm.yWorld + extentsNm.cyWorld, dyMaxNm) ;
        }
      DBG_P (fprintf (stderr, "Layer \"%s\" extents:(%lf,%lf)<->(%lf,%lf)=<%lfx%lf>\n", ((LAYER *)(lstItr->data))->pszDescription,
        dxTopNm, dyTopNm, dxBotNm, dyBotNm, dxBotNm - dxTopNm, dyBotNm - dyTopNm)) ;
      }
    }

  cxNm = dxMaxNm - dxMinNm ;
  cyNm = dyMaxNm - dyMinNm ;

  DBG_P (fprintf (stderr, "Design extents: (%lf,%lf)<%lfx%lf>\n", dxMinNm, dyMinNm, cxNm, cyNm)) ;

  // If we are to fit to n x m pages instead of using a fixed scale, let's re-calculate the scale here,
  // since the GtkAdjustment approximates the scale to as many decimals as specified
  if (pPO->bFit)
    pPO->dPointsPerNano = MIN (dEffPageCXPts * pPO->iCXPages / cxNm,
                               dEffPageCYPts * pPO->iCYPages / cyNm) ;

  if (pPO->bCenter)
    {
    double dWidthPoints = cxNm * pPO->dPointsPerNano,
           dHeightPoints = cyNm * pPO->dPointsPerNano ;
    dxDiffMinNm = (((dEffPageCXPts * pPO->iCXPages) - dWidthPoints) / 2) / pPO->dPointsPerNano ;
    dyDiffMinNm = (((dEffPageCYPts * pPO->iCYPages) - dHeightPoints) / 2) / pPO->dPointsPerNano ;
    }

  lstPages = PlaceObjectsOnPages (design, pPO,
    dEffPageCXPts / pPO->dPointsPerNano, dEffPageCYPts / pPO->dPointsPerNano,
    dxMinNm - dxDiffMinNm, dyMinNm - dyDiffMinNm) ;

  if (NULL == (pfile = OpenPrintStream ((print_OP *)pPO)))
    {
    fprintf (stderr, "Failed to open print stream \"%s\"\n", pPO->po.pszPrintString) ;
    return ;
    }

  PrintMagic (pfile, pPO->po.bPortrait, pPO->po.dPaperCX, pPO->po.dPaperCY) ;

  fprintf (pfile,
    "%%%%BeginProlog\n"
    "/epsilon 0.1 def\n"
    "/nm { %s mul } def\n"
    "/labelfontsize 12 nm def\n"
    "1 setlinejoin 1 setlinecap\n"
    "\n"
    "%% a b min\n"
    "/min\n"
    "  {\n"
    "  /b exch def\n"
    "  /a exch def\n"
    "\n"
    "  a b lt { a } { b } ifelse\n"
    "  } def\n"
    "\n"
    "%% a b max\n"
    "/max\n"
    "  {\n"
    "  /b exch def\n"
    "  /a exch def\n"
    "\n"
    "  a b gt { a } { b } ifelse\n"
    "  } def\n"
    "\n"
    "/linewidth 0.1 nm 1 min def\n"
    "\n"
    PS_TEXT_PLACEMENT_PREAMBLE
    "linewidth setlinewidth\n",
    g_ascii_dtostr (doubles[0], G_ASCII_DTOSTR_BUF_SIZE, pPO->dPointsPerNano)) ;

  PrintObjectPreamble (pfile, lstPages, pPO->iCXPages * pPO->iCYPages) ;

  fprintf (pfile, "%%%%EndProlog\n") ;

  PrintPages (pfile, pPO, lstPages, dEffPageCYPts, dxMinNm, dyMinNm, dxDiffMinNm, dyDiffMinNm) ;

  PrintTrailer (pfile, pPO->iCXPages * pPO->iCYPages) ;

  ClosePrintStream (pfile, (print_OP *)pPO) ;
  }

static GList **PlaceObjectsOnPages (DESIGN *design, print_design_OP *pPO, double cxPageNm, double cyPageNm, double dxMinNm, double dyMinNm)
  {
  gboolean *pbPrint = NULL, bPrintThisLayer = FALSE ;
  GList
    **lstRet = g_malloc0 (pPO->iCXPages * pPO->iCYPages * sizeof (GList *)),
    *lstItr = NULL,
    *lstItrObjs = NULL ;
  int idxX1 = -1, idxX2 = -1, idxY1 = -1, idxY2 = -1, Nix1, Nix2, idx = -1 ;
  WorldRectangle rcWorld ;

  DBG_P (fprintf (stderr, "PlaceObjectsOnPage:Entering\n")) ;

  for (lstItr = design->lstLayers ; lstItr != NULL ; lstItr = lstItr->next)
    {
    DBG_P (fprintf (stderr, "PlaceObjectsOnPage:Checking whether to examine layer \"%s\"...", ((LAYER *)(lstItr->data))->pszDescription)) ;
    if (!(bPrintThisLayer = (NULL == (pbPrint = g_object_get_data (G_OBJECT (lstItr->data), PRINT_LAYER_KEY)))))
      bPrintThisLayer = (*pbPrint) ;

    DBG_P (fprintf (stderr, "%s\n", bPrintThisLayer ? "yes" : "no")) ;

    if (bPrintThisLayer)
      {
      for (lstItrObjs = (QCAD_LAYER (lstItr->data))->lstObjs ; lstItrObjs != NULL ; lstItrObjs = lstItrObjs->next)
        if (NULL != lstItrObjs->data)
          {
          qcad_design_object_get_bounds_box (QCAD_DESIGN_OBJECT (lstItrObjs->data), &rcWorld) ;

          DBG_P (fprintf (stderr, "Found bounds box:(%lf,%lf)<%lfx%lf>\n", x, y, cx, cy)) ;

          idxX1 = (int) ((rcWorld.xWorld - dxMinNm) / cxPageNm) ;
          idxX1 = idxX1 < 0 ? 0 : idxX1 >= pPO->iCXPages ? pPO->iCXPages - 1 : idxX1 ; // Paranoia check
          idxY1 = (int) ((rcWorld.yWorld - dyMinNm) / cyPageNm) ;
          idxY1 = idxY1 < 0 ? 0 : idxY1 >= pPO->iCYPages ? pPO->iCYPages - 1 : idxY1 ; // Paranoia check
          idxX2 = (int) ((rcWorld.xWorld + rcWorld.cxWorld - dxMinNm) / cxPageNm) ;
          idxX2 = idxX2 < 0 ? 0 : idxX2 >= pPO->iCXPages ? pPO->iCXPages - 1 : idxX2 ; // Paranoia check
          idxY2 = (int) ((rcWorld.yWorld + rcWorld.cyWorld - dyMinNm) / cyPageNm) ;
          idxY2 = idxY2 < 0 ? 0 : idxY2 >= pPO->iCYPages ? pPO->iCYPages - 1 : idxY2 ; // Paranoia check

          for (Nix1 = idxX1 ; Nix1 <= idxX2 ; Nix1++)
            for (Nix2 = idxY1 ; Nix2 <= idxY2 ; Nix2++)
              {
              DBG_P (fprintf (stderr, "Adding an object of type %s to page (%d,%d)\n",
                g_type_name (G_TYPE_FROM_INSTANCE (lstItrObjs->data)), Nix1, Nix2)) ;
              idx = pPO->bPrintOrderOver ? Nix2 * pPO->iCXPages + Nix1 : Nix1 * pPO->iCYPages + Nix2 ;
              lstRet[idx] = g_list_prepend (lstRet[idx], lstItrObjs->data) ;
              }
          }
      }
    }

  return lstRet ;
  }

static void PrintObjectPreamble (FILE *pfile, GList **lstPages, int icPages)
  {
  int Nix ;
  GList *lstObjs = NULL, *lstItr = NULL ;

  // For each type of object appearing on the pages, add a reference to the linked list lstObjs
  for (Nix = 0 ; Nix < icPages ; Nix++)
    for (lstItr = lstPages[Nix] ; lstItr != NULL ; lstItr = lstItr->next)
      lstObjs = qcad_design_object_add_types (QCAD_DESIGN_OBJECT (lstItr->data), lstObjs) ;

  // For each link in lstObjs, print out the preamble
  for (lstItr = lstObjs ; lstItr != NULL ; lstItr = lstItr->next)
    fprintf (pfile, "\n%s", 
      qcad_design_object_class_get_PostScript_preamble (QCAD_DESIGN_OBJECT_CLASS (g_type_class_peek ((GType)(lstItr->data))))) ;
  fprintf (pfile, "\n") ;
  }

void PrintPages (FILE *pfile, print_design_OP *pPO, GList **lstPages, double dEffPageCYPts, double dxMinNm, double dyMinNm, double dxDiffMinNm, double dyDiffMinNm)
  {
  int icPages = pPO->iCXPages * pPO->iCYPages, Nix ;
  GList *lstItr = NULL ;
  char *pszPSInstance = NULL ;
  double cxPageOffsetNm = 0, cyPageOffsetNm = 0 ;
  char doubles[19][G_ASCII_DTOSTR_BUF_SIZE] = {""} ;

  for (Nix = 0 ; Nix < icPages ; Nix++)
    {
    DBG_P (fprintf (stderr, "PrintPages:printing page %d: cxPageOffsetNm = %lf cyPageOffsetNm = %lf\n",
      Nix, cxPageOffsetNm, cyPageOffsetNm)) ;
//
//  +->-+ Margins are drawn like this (because the PostScript origin is bottom left)
//  |   |
//  ^   v
//  |   |
//  +-<-+
//  ^
//  |
//
    fprintf (pfile,
      "%%%%Page: %d %d\n"
      "%%%%BeginPageSetup\n"
      "/nmx { %s sub %s sub %s add 1 nm mul %s add } def\n"
      "/nmy { %s exch %s sub %s sub %s add 1 nm mul %s add sub } def\n"
      "%%%%EndPageSetup\n"
      "gsave\n"
      "%%The margin\n"
      "newpath\n"
      "%s %s moveto\n"
      "%s %s lineto\n"
      "%s %s lineto\n"
      "%s %s lineto\n"
      "closepath eoclip\n",
      Nix + 1, 
      Nix + 1, // %%Page:
      g_ascii_dtostr (doubles[ 0], G_ASCII_DTOSTR_BUF_SIZE, dxMinNm), 
      g_ascii_dtostr (doubles[ 1], G_ASCII_DTOSTR_BUF_SIZE, cxPageOffsetNm), 
      g_ascii_dtostr (doubles[ 2], G_ASCII_DTOSTR_BUF_SIZE, dxDiffMinNm), 
      g_ascii_dtostr (doubles[ 4], G_ASCII_DTOSTR_BUF_SIZE, pPO->po.dLMargin), // nmx
      g_ascii_dtostr (doubles[ 5], G_ASCII_DTOSTR_BUF_SIZE, pPO->po.dPaperCY), 
      g_ascii_dtostr (doubles[ 6], G_ASCII_DTOSTR_BUF_SIZE, dyMinNm), 
      g_ascii_dtostr (doubles[ 7], G_ASCII_DTOSTR_BUF_SIZE, cyPageOffsetNm), 
      g_ascii_dtostr (doubles[ 8], G_ASCII_DTOSTR_BUF_SIZE, dyDiffMinNm), 
      g_ascii_dtostr (doubles[10], G_ASCII_DTOSTR_BUF_SIZE, pPO->po.dTMargin), // nmy
      g_ascii_dtostr (doubles[11], G_ASCII_DTOSTR_BUF_SIZE, pPO->po.dLMargin), 
      g_ascii_dtostr (doubles[12], G_ASCII_DTOSTR_BUF_SIZE, pPO->po.dBMargin), // moveto
      g_ascii_dtostr (doubles[13], G_ASCII_DTOSTR_BUF_SIZE, pPO->po.dLMargin), 
      g_ascii_dtostr (doubles[14], G_ASCII_DTOSTR_BUF_SIZE, pPO->po.dPaperCY - pPO->po.dTMargin), // lineto
      g_ascii_dtostr (doubles[15], G_ASCII_DTOSTR_BUF_SIZE, pPO->po.dPaperCX - pPO->po.dRMargin), 
      g_ascii_dtostr (doubles[16], G_ASCII_DTOSTR_BUF_SIZE, pPO->po.dPaperCY - pPO->po.dTMargin), // lineto
      g_ascii_dtostr (doubles[17], G_ASCII_DTOSTR_BUF_SIZE, pPO->po.dPaperCX - pPO->po.dRMargin), 
      g_ascii_dtostr (doubles[18], G_ASCII_DTOSTR_BUF_SIZE, pPO->po.dBMargin) // lineto
      ) ;

    if (pPO->bPrintOrderOver)
      cxPageOffsetNm += (pPO->po.dPaperCX - pPO->po.dLMargin - pPO->po.dRMargin) / pPO->dPointsPerNano ;
    else
      cyPageOffsetNm += (pPO->po.dPaperCY - pPO->po.dTMargin - pPO->po.dBMargin) / pPO->dPointsPerNano ;

    if (pPO->bPrintOrderOver && !((Nix + 1) % pPO->iCXPages))
      {
      DBG_P (fprintf (stderr,  "PrintPages:bPrintOrderOver:Increasing cy offset\n")) ;
      cyPageOffsetNm += (pPO->po.dPaperCY - pPO->po.dTMargin - pPO->po.dBMargin) / pPO->dPointsPerNano ;
      cxPageOffsetNm = 0 ;
      }
    else
    if (!pPO->bPrintOrderOver && !((Nix + 1) % pPO->iCYPages))
      {
      DBG_P (fprintf (stderr,  "PrintPages:!bPrintOrderOver:Increasing cx offset\n")) ;
      cxPageOffsetNm += (pPO->po.dPaperCX - pPO->po.dLMargin - pPO->po.dRMargin) / pPO->dPointsPerNano ;
      cyPageOffsetNm = 0 ;
      }

    for (lstItr = lstPages[Nix] ; lstItr != NULL ; lstItr = lstItr->next)
      {
      pszPSInstance = qcad_design_object_get_PostScript_instance (QCAD_DESIGN_OBJECT (lstItr->data), pPO->po.bPrintColours) ;
      fprintf (pfile, "%s\n", pszPSInstance) ;
      g_free (pszPSInstance) ;
      }

    fprintf (pfile,
      "grestore\n"
      "showpage\n"
      "%%%%PageTrailer\n") ;
    }
  }
